pkg/actors/internal/placement/placement.go

```diff
@@ -0,0 +1,360 @@
+	Healthz   healthz.Healthz
+}
+
+type placement struct {
+	client     *client.Client
+	actorTable table.Interface
+	reminders  storage.Interface
+	apiLevel   *apilevel.APILevel
+	htarget    healthz.Target
+
+	hashTable         *hashing.ConsistentHashTables
+	virtualNodesCache *hashing.VirtualNodesCache
+
+	lock          *lock.Lock
+	lockVersion   atomic.Uint64
+	updateVersion atomic.Uint64
+	operationLock *fifo.Mutex
+
+	appID     string
+	namespace string
+	hostname  string
+	port      string
+	isReady   atomic.Bool
+	readyCh   chan struct{}
+	wg        sync.WaitGroup
+	closed    atomic.Bool
+}
+
+func New(opts Options) (Interface, error) {
+	lock := lock.New()
+	client, err := client.New(client.Options{
+		Addresses: opts.Addresses,
+		Security:  opts.Security,
+		Table:     opts.Table,
+		Lock:      lock,
+		Healthz:   opts.Healthz,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &placement{
+		client:            client,
+		actorTable:        opts.Table,
+		virtualNodesCache: hashing.NewVirtualNodesCache(),
+		hashTable: &hashing.ConsistentHashTables{
+			Entries: make(map[string]*hashing.Consistent),
+		},
+		reminders:     opts.Reminders,
+		appID:         opts.AppID,
+		port:          strconv.Itoa(opts.Port),
+		namespace:     opts.Namespace,
+		hostname:      opts.Hostname,
+		operationLock: fifo.New(),
+		apiLevel:      opts.APILevel,
+		htarget:       opts.Healthz.AddTarget(),
+		lock:          lock,
+		readyCh:       make(chan struct{}),
+	}, nil
+}
+
+func (p *placement) Run(ctx context.Context) error {
+	err := concurrency.NewRunnerManager(
+		p.client.Run,
+		func(ctx context.Context) error {
+			ch, actorTypes := p.actorTable.SubscribeToTypeUpdates(ctx)
+			log.Infof("Reporting actor types: %v", actorTypes)
+			if err := p.sendHost(ctx, actorTypes); err != nil {
+				return err
+			}
+
+			tch := time.NewTicker(statusReportHeartbeatInterval)
+			defer tch.Stop()
+
+			for {
+				select {
+				case actorTypes = <-ch:
+					log.Infof("Updating actor types: %v", actorTypes)
+					// TODO: @joshvanl: it is a nonsense to be sending the same actor
+					// types over and over again, *every second*. This should be removed
+					// ASAP.
+				case <-tch.C:
+				case <-ctx.Done():
+					return nil
+				}
+
+				if err := p.sendHost(ctx, actorTypes); err != nil {
+					return err
+				}
+			}
+		},
+		func(ctx context.Context) error {
+			defer p.wg.Wait()
+			defer p.lock.EnsureUnlockTable()
+
+			for {
+				in, err := p.client.Recv(ctx)
+				if err != nil {
+					return err
+				}
+
+				p.handleReceive(ctx, in)
+			}
+		},
+	).Run(ctx)
+
+	p.closed.Store(true)
+	p.lock.EnsureUnlockTable()
+	return err
+}
+
+// LookupActor returns the address of the actor.
+// Placement _must_ be locked before calling this method.
+func (p *placement) LookupActor(ctx context.Context, req *api.LookupActorRequest) (*api.LookupActorResponse, error) {
+	table, ok := p.hashTable.Entries[req.ActorType]
+	if !ok {
+		return nil, ErrActorTypeNotFound
```

Comment:
```suggestion
		return nil, messages.ErrActorNoAddress
```
Related to comment on line 44